galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker_test.go

```diff
@@ -0,0 +1,392 @@
+				policy: `
+targets:
+- name: foobar
+  ports:
+  - number: 8080
+peers:
+- mtls:
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    false,
+		},
+		"policy matches service but is not strict": {
+			policy: PolicyResource{
+
+				namespace: "my-namespace",
+				policy: `
+targets:
+- name: foobar
+  ports:
+  - number: 8080
+peers:
+- mtls:
+    mode: PERMISSIVE
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    false,
+		},
+		"policy matches service but uses deprecated field": {
+			policy: PolicyResource{
+
+				namespace: "my-namespace",
+				policy: `
+targets:
+- name: foobar
+  ports:
+  - number: 8080
+peers:
+- mtls:
+    allowTls: true
+    mode: STRICT
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    false,
+		},
+		"policy matches service but peer is optional": {
+			policy: PolicyResource{
+
+				namespace: "my-namespace",
+				policy: `
+targets:
+- name: foobar
+  ports:
+  - number: 8080
+peers:
+- mtls:
+    mode: STRICT
+peerIsOptional: true
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    false,
+		},
+		"policy matches service but does not use mtls": {
+			policy: PolicyResource{
+
+				namespace: "my-namespace",
+				policy: `
+targets:
+- name: foobar
+  ports:
+  - number: 8080
+peers:
+- jwt: # undocumented setting?
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    false,
+		},
+		"policy matches every port on service": {
+			policy: PolicyResource{
+
+				namespace: "my-namespace",
+				policy: `
+targets:
+- name: foobar
+peers:
+- mtls:
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    true,
+		},
+		"policy matches every service in namespace": {
+			policy: PolicyResource{
+				namespace: "my-namespace",
+				policy: `
+peers:
+- mtls:
+`,
+			},
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    true,
+		},
+		"policy matches entire mesh": {
+			meshPolicy: `
+peers:
+- mtls:
+`,
+			service: NewTargetServiceWithPortNumber("foobar.my-namespace.svc.cluster.local", 8080),
+			want:    true,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			pc := NewPolicyChecker()
+
+			if tc.meshPolicy != "" {
+				meshpb, err := yAMLToPolicy(tc.meshPolicy)
+				if err != nil {
+					t.Fatalf("expected: %v, got error when parsing yaml: %v", tc.want, err)
+				}
+				pc.AddMeshPolicy(meshpb)
+			}
+
+			if tc.policy.policy != "" {
+				pb, err := yAMLToPolicy(tc.policy.policy)
+				if err != nil {
+					t.Fatalf("expected: %v, got error when parsing yaml: %v", tc.want, err)
+				}
+				pc.AddPolicy(tc.policy.namespace, pb)
```

Comment:
Tests should check the error return here. (I don't mind if you end up adding this in followup work).

Reply:
I missed this (previously AddPolicy did not return)!

I remembered that `errcheck` (now integrated into `glangci-lint`) can detect this:
```
~/g/s/i/istio (add_mtls_policy_checker|✔) [1] $ ~/go/bin/golangci-lint run galley/pkg/config/analysis/analyzers/auth/mtls/
galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker_test.go:246:17: Error return value of `pc.AddPolicy` is not checked (errcheck)
                                pc.AddPolicy(tc.policy.namespace, pb)
                                            ^
galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker_test.go:366:17: Error return value of `pc.AddPolicy` is not checked (errcheck)
                                pc.AddPolicy(p.namespace, pb)
```
Any idea if we run this as part of the lint check in istio?

Reply:
(the lint check refuses to run on os x due to xargs incompatibilities, and i haven't had a chance to see if it could be made to work)

Reply:
My guess is that the CI linting is configured to skip over tests, or is more lenient with them.